src/main/java/org/opentripplanner/routing/algorithm/via/ViaRoutingWorker.java

```diff
@@ -0,0 +1,152 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import org.opentripplanner.model.plan.Itinerary;
+import org.opentripplanner.routing.api.request.RouteRequest;
+import org.opentripplanner.routing.api.request.request.RouteViaRequest;
+import org.opentripplanner.routing.api.request.request.ViaLocation;
+import org.opentripplanner.routing.api.response.InputField;
+import org.opentripplanner.routing.api.response.RoutingError;
+import org.opentripplanner.routing.api.response.RoutingErrorCode;
+import org.opentripplanner.routing.api.response.RoutingResponse;
+import org.opentripplanner.routing.api.response.ViaRoutingResponse;
+import org.opentripplanner.routing.error.RoutingValidationException;
+
+public class ViaRoutingWorker {
+
+  private final RouteViaRequest viaRequest;
+  private final RouteRequest request;
+  private final Function<RouteRequest, RoutingResponse> routingWorker;
+
+  public ViaRoutingWorker(
+    RouteViaRequest request,
+    Function<RouteRequest, RoutingResponse> routingWorker
+  ) {
+    this.request = request.routeRequest().clone();
+    this.viaRequest = request;
+    this.routingWorker = routingWorker;
+  }
+
+  public ViaRoutingResponse route() {
+    //Loop over Via, for each cycle change from/to and JourneyRequest.
+    var result = viaRequest.viaLegs().stream().map(v -> getRoutingResponse(viaRequest, v)).toList();
+
+    return combineRoutingResponse(result);
+  }
+
+  /**
+   * Set to point and search trips. Return result with itineraries and prepare request for next
+   * search.
+   */
+  private RoutingResponse getRoutingResponse(RouteViaRequest request, RouteViaRequest.ViaLeg v) {
+    // If viaLocation is null then we are at last search
+    if (v.viaLocation() != null) {
+      this.request.setTo(v.viaLocation().point());
+    } else {
+      this.request.setTo(request.routeRequest().to());
+    }
+
+    var e = new RoutingError(
+      RoutingErrorCode.NO_TRANSIT_CONNECTION_IN_SEARCH_WINDOW,
+      InputField.INTERMEDIATE_PLACE
+    );
+    var response = this.routingWorker.apply(this.request);
+    var firstArrival = firstArrival(response)
+      .orElseThrow(() -> new RoutingValidationException(List.of(e)));
+    var lastArrival = lastArrival(response)
+      .orElseThrow(() -> new RoutingValidationException(List.of(e)));
+    var maxSlack = Optional
+      .ofNullable(v.viaLocation())
+      .map(ViaLocation::maxSlack)
+      .orElse(ViaLocation.DEFAULT_MAX_SLACK);
+
+    if (v.viaLocation() == null) {
+      return response;
+    }
+
+    // Prepare next search
+    var searchWindow = Duration.between(firstArrival, lastArrival).plus(maxSlack);
+
+    this.request.setSearchWindow(searchWindow);
+    this.request.setDateTime(firstArrival.plus(v.viaLocation().minSlack()).toInstant());
+    this.request.setFrom(v.viaLocation().point());
+    this.request.setJourney(v.journeyRequest());
```

Comment:
Wee need to set numItineraries here to unlimited. Otherwise we don't get any connecting services for the later itineraries if going from a low frequency area to a high frequency area

Reply:
It needs to be set from the request for the first search, but not for the subsequent ones. However we might want to re-use the value also in `combineRoutingResponse`, so that we have a maximum number of combinations per itinerary.